ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3PutHandler.java

```diff
@@ -93,20 +96,42 @@ protected void doHandle(RestRequest restRequest, RestResponseChannel restRespons
     // Overwrites are allowed by default for S3 APIs https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html
     restRequest.setArg(NAMED_UPSERT, true);
 
+    // Calculate checksum of the file content. This will be used in ETag header. Default algorithm seems to be MD5
+    try {
+      restRequest.setDigestAlgorithm("MD5");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RestServiceException("MD5 Algorithm is not supported", e, RestServiceErrorCode.InternalServerError);
+    }
+
     // 2. Upload the blob by following named blob PUT path
     namedBlobPutHandler.handle(restRequest, restResponseChannel, buildCallback(metrics.s3PutHandleMetrics, (r) -> {
       if (restResponseChannel.getStatus() == ResponseStatus.Created) {
         // Set the response status to 200 since Ambry named blob PUT has response as 201.
         restResponseChannel.setStatus(ResponseStatus.Ok);
         // Set S3 ETag header
-        String blobId = (String) restResponseChannel.getHeader(LOCATION);
-        blobId = blobId.startsWith("/") ? blobId.substring(1) : blobId;
-        restResponseChannel.setHeader("ETag", blobId);
+        // TODO: For Multipart uploads, ETag is calculated as md5(part1) + md5(part2) + ... + md5(partN) + "-" + N.
+        //  This needs some external db to store md5 of individual parts.
+        String eTag = calculateETag(restRequest);
+        restResponseChannel.setHeader("ETag", eTag);
       }
       callback.onCompletion(null, null);
     }, restRequest.getUri(), LOGGER, (result, exception) -> {
       LOGGER.info("Request {} failed in s3 put object API", restRequest.getUri(), exception);
       callback.onCompletion(null, exception);
     }));
   }
+
+  private String calculateETag(RestRequest restRequest) {
+    byte[] digest = restRequest.getDigest();
+    StringBuilder hexString = new StringBuilder();
+    for (byte b : digest) {
+      String hex = Integer.toHexString(0xff & b);
+      if (hex.length() == 1) {
+        hexString.append('0');
+      }
+      hexString.append(hex);
+    }
```

Comment:
can we create a helper function in Utils class and move this logic over there, a method like encodeToHexString?